README


This archive contains the distribution Log-Dispatch-File-Stamped,
version 0.15:

  Logging to date/time stamped files

This software is copyright (c) 2003 by Karen Etheridge.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.023.



Changes
Revision history for Log-Dispatch-File-Stamped

0.15      2014-11-05 02:28:41Z
          - fix tests on MSWin32 using arch-sensitive file separators

0.14      2014-11-03 06:14:34Z
          - fix lingering Test::TempDir prereq
          - fix remaining issues with running tests in parallel by using
            unique tempdirs for all tests

0.13      2014-02-17 01:14:47Z
          - uses of Test::TempDir switched to Path::Tiny

0.12      2013-06-17 21:57:03Z
          - conversion to Dist::Zilla

0.11      2013-06-17
          - re-release as a stable release

0.10_010  2013-05-04
          - fixed testing of octal values on MSWin32, ONCE AND FOR ALL

0.10_009  2013-05-01
          - even more fixing of testing of octal values on MSWin32

0.10_008  2013-04-30
          - more fixing of testing of octal values on MSWin32

0.10_007  2013-04-29
          - fix testing of octal values on MSWin32

0.10_003  2013-04-22
          - fix version declaration (Karen Etheridge)

0.10_002  2013-04-22
          - fix version requirements (Karen Etheridge)

0.10_001  2013-04-09
          - add support for all remaining Log::Dispatch::File options (Karen Etheridge)

0.10  2011-09-25
          - add support for permissions parameter - reported by Jonathan Peyser

0.09  2010-08-15
          - fix incorrect number of skipped tests under perl < 5.8

0.08  2010-08-14
          - skip binmode tests under perls < 5.8

0.07  2010-08-13
          - add support for binmode parameter - patch by Tokuhiro Matsuno
            (https://rt.cpan.org/Ticket/Display.html?id=60278)

0.06  2008-06-29
          - added missing meta-spec to meta.yml
          - remove pod tests

0.05  2006-09-07
          - fix typo which prevented correcting indexing by PAUSE
          - move .pm file to lib/
          - document log_message()

0.04  2006-09-07
          - add Build.PL to build with Module::Build
          - add t/podcover.t to make CPANTS happy

0.03  2004-09-15
          - Fix "Log::Dispatch::* object file already exists" warning
            displayed on some platforms during "make test".
          - Require Test::More
          - Fix pod and add a pod test

0.02  2004-08-08
          - log_to() wasn't working due to attribute name clash
            (https://rt.cpan.org/Ticket/Display.html?id=7249)

0.01  2003-02-08
          - Uploaded 0.01 to CPAN



Makefile.PL
# This Makefile.PL for Log-Dispatch-File-Stamped was generated by
# Dist::Zilla::Plugin::MakeMaker::Fallback 0.014
# and Dist::Zilla::Plugin::MakeMaker::Awesome 0.31.
# Don't edit it but the dist.ini and plugins used to construct it.

use strict;
use warnings;

BEGIN {
my %configure_requires = (
    'ExtUtils::MakeMaker' => '0',
    'Module::Build::Tiny' => '0.039',
);

my @missing = grep {
    ! eval "require $_; $_->VERSION($configure_requires{$_}); 1"
} keys %configure_requires;

if (not @missing)
{
    print "Congratulations, your toolchain understands 'configure_requires'!\n\n";
}
else
{
    $ENV{PERL_MM_FALLBACK_SILENCE_WARNING} or warn <<'EOW';
*** WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING ***

If you're seeing this warning, your toolchain is really, really old* and you'll
almost certainly have problems installing CPAN modules from this century. But
never fear, dear user, for we have the technology to fix this!

If you're using CPAN.pm to install things, then you can upgrade it using:

    cpan CPAN

If you're using CPANPLUS to install things, then you can upgrade it using:

    cpanp CPANPLUS

If you're using cpanminus, you shouldn't be seeing this message in the first
place, so please file an issue on github.

If you're installing manually, please retrain your fingers to run Build.PL
when present instead.

This public service announcement was brought to you by the Perl Toolchain
Gang, the irc.perl.org #toolchain IRC channel, and the number 42.

----

* Alternatively, you are doing something overly clever, in which case you
should consider setting the 'prefer_installer' config option in CPAN.pm, or
'prefer_makefile' in CPANPLUS, to 'mb" and '0' respectively.

You can also silence this warning for future installations by setting the
PERL_MM_FALLBACK_SILENCE_WARNING environment variable.

EOW

    sleep 10 if -t STDIN && (-t STDOUT || !(-f STDOUT || -c STDOUT));
}
}

use 5.006;
use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Logging to date/time stamped files",
  "AUTHOR" => "Eric Cholet <cholet\@logilune.com>, Karen Etheridge <ether\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "Module::Build::Tiny" => "0.039"
  },
  "DISTNAME" => "Log-Dispatch-File-Stamped",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Log::Dispatch::File::Stamped",
  "PREREQ_PM" => {
    "File::Basename" => 0,
    "File::Spec::Functions" => 0,
    "Log::Dispatch::File" => "2.38",
    "POSIX" => 0,
    "Params::Validate" => 0,
    "namespace::clean" => 0,
    "parent" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Log::Dispatch" => 0,
    "Path::Tiny" => 0,
    "Test::Deep" => 0,
    "Test::More" => "0.88"
  },
  "VERSION" => "0.15",
  "test" => {
    "TESTS" => "t/*.t"
  }
);

my %FallbackPrereqs = (
  "ExtUtils::MakeMaker" => 0,
  "File::Basename" => 0,
  "File::Spec" => 0,
  "File::Spec::Functions" => 0,
  "Log::Dispatch" => 0,
  "Log::Dispatch::File" => "2.38",
  "Module::Build::Tiny" => "0.039",
  "POSIX" => 0,
  "Params::Validate" => 0,
  "Path::Tiny" => 0,
  "Test::Deep" => 0,
  "Test::More" => "0.88",
  "namespace::clean" => 0,
  "parent" => 0,
  "strict" => 0,
  "warnings" => 0
);

unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/Log/Dispatch/File/Stamped.pm
use strict;
use warnings;
package Log::Dispatch::File::Stamped;
# git description: v0.14-1-g1c79ca2
$Log::Dispatch::File::Stamped::VERSION = '0.15';
# ABSTRACT: Logging to date/time stamped files
# KEYWORDS: log logging output file timestamp date rolling rotate

use File::Basename        qw(fileparse);
use File::Spec::Functions qw(catfile);
use POSIX                 qw(strftime);

use Log::Dispatch::File 2.38;
use parent 'Log::Dispatch::File';

use Params::Validate qw(validate SCALAR);
Params::Validate::validation_options( allow_extra => 1 );

use namespace::clean;


sub _basic_init
{
    my $self = shift;

    $self->SUPER::_basic_init(@_);

    my %p = validate(
        @_,
        {
            stamp_fmt => {
                type => SCALAR,
                default => '%Y%m%d',
            },
        },
    );

    $self->{stamp_fmt} = $p{stamp_fmt};

    # cache of last timestamp used
    $self->{_stamp} = '';

    # split pathname into path, basename, extension
    @$self{qw(_name _path _ext)} = fileparse($self->{filename}, '\.[^.]+');

    # stored in $self->{filename} (overwrites original); used by _open_file()
    $self->_make_filename;
}

sub _make_filename
{
    my $self = shift;

    my $stamp = strftime($self->{stamp_fmt}, localtime);

    # re-use last filename if the stamp has not changed
    return $self->{filename} if $stamp eq $self->{_stamp};

    # build the stamped file name
    my $filename = join '-', $self->{_name}, $stamp;
    $filename .= $self->{_ext} if $self->{_ext};
    $self->{filename} = catfile($self->{_path}, $filename);
}

sub log_message
{
    my $self = shift;

    # check if the filename is the same as last time...
    my $old_filename = $self->{filename};
    $self->_make_filename;

    # don't re-open if we use close-after-write - the superclass will do it
    if (not $self->{close} and $old_filename ne $self->{filename})
    {
        $self->_open_file;
    }

    $self->SUPER::log_message(@_);
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Log::Dispatch::File::Stamped - Logging to date/time stamped files

=head1 VERSION

version 0.15

=head1 SYNOPSIS

  use Log::Dispatch::File::Stamped;

  my $file = Log::Dispatch::File::Stamped->new(
    name      => 'file1',
    min_level => 'info',
    filename  => 'Somefile.log',
    stamp_fmt => '%Y%m%d',
    mode      => 'append' );

  $file->log( level => 'emerg', message => "I've fallen and I can't get up\n" );

=head1 DESCRIPTION

This module subclasses L<Log::Dispatch::File> for logging to date/time
stamped files, respecting all its configuration options.

=head1 METHODS

=head2 new(%p)

This method takes the same set of parameters as L<Log::Dispatch::File::new()|Log::Dispatch::File/new>,
with the following differences:

=over 4

=item * filename ($)

The filename template. The actual timestamp will be appended to this filename
when creating the actual logfile. If the filename has an extension, the
timestamp is inserted before the extension. See examples below.

=item * stamp_fmt ($)

The format of the timestamp string. This module uses L<POSIX::strftime|POSIX/strftime> to
create the timestamp string from the current local date and time.
Refer to your platform's C<strftime> documentation for the list of allowed
tokens.

Defaults to C<%Y%m%d>.

=back

=head2 log_message( message => $ )

Sends a message to the appropriate output.  Generally this
shouldn't be called directly but should be called through the
C<log()> method (in L<Log::Dispatch::Output>).

=head1 EXAMPLES

=for stopwords txt

Assuming the current date and time is:

  % perl -e 'print scalar localtime'
  Sat Feb  8 13:56:13 2003

  Log::Dispatch::File::Stamped->new(
    name      => 'file',
    min_level => 'debug',
    filename  => 'logfile.txt',
  );

This will log to file F<logfile-20030208.txt>.

  Log::Dispatch::File::Stamped->new(
    name      => 'file',
    min_level => 'debug',
    filename  => 'logfile.txt',
    stamp_fmt => '%d%H',
  );

This will log to file F<logfile-0813.txt>.

=head1 SEE ALSO

L<Log::Dispatch::File>, L<POSIX>.

=head1 ACKNOWLEDGEMENTS

=for stopwords Rolsky

Dave Rolsky, author of the Log::Dispatch suite and many other
fine modules on CPAN.

This module was rewritten to respect all present (and future) options to
L<Log::Dispatch::File> by Karen Etheridge, <ether@cpan.org>.

=head1 AUTHORS

=over 4

=item *

Eric Cholet <cholet@logilune.com>

=item *

Karen Etheridge <ether@cpan.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2003 by Karen Etheridge.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Build.PL
# This Build.PL for Log-Dispatch-File-Stamped was generated by
# Dist::Zilla::Plugin::ModuleBuildTiny::Fallback 0.010
use strict;
use warnings;

if (eval 'use Module::Build::Tiny 0.039; 1')
{
    print "Congratulations, your toolchain understands 'configure_requires'!\n\n";

    # This section for Log-Dispatch-File-Stamped was generated by Dist::Zilla::Plugin::ModuleBuildTiny 0.008.
    use strict;
    use warnings;

    use 5.006;
    # use Module::Build::Tiny 0.039;
    Build_PL();
}
else
{
    $ENV{PERL_MB_FALLBACK_SILENCE_WARNING} or warn <<'EOW';
*** WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING ***

If you're seeing this warning, your toolchain is really, really old* and you'll
almost certainly have problems installing CPAN modules from this century. But
never fear, dear user, for we have the technology to fix this!

If you're using CPAN.pm to install things, then you can upgrade it using:

    cpan CPAN

If you're using CPANPLUS to install things, then you can upgrade it using:

    cpanp CPANPLUS

If you're using cpanminus, you shouldn't be seeing this message in the first
place, so please file an issue on github.

This public service announcement was brought to you by the Perl Toolchain
Gang, the irc.perl.org #toolchain IRC channel, and the number 42.

----

* Alternatively, you are running this file manually, in which case you need
to learn to first fulfill all configure requires prerequisites listed in
META.yml or META.json -- or use a cpan client to install this distribution.

You can also silence this warning for future installations by setting the
PERL_MB_FALLBACK_SILENCE_WARNING environment variable, but please don't do
that until you fix your toolchain as described above.

EOW
    sleep 10 if -t STDIN && (-t STDOUT || !(-f STDOUT || -c STDOUT));


    # This section was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.023.
    use strict;
    use warnings;

    require Module::Build; Module::Build->VERSION(0.28);


    my %module_build_args = (
      "build_requires" => {},
      "configure_requires" => {
        "ExtUtils::MakeMaker" => 0,
        "Module::Build::Tiny" => "0.039",
        "perl" => "5.006"
      },
      "dist_abstract" => "Logging to date/time stamped files",
      "dist_author" => [
        "Eric Cholet <cholet\@logilune.com>",
        "Karen Etheridge <ether\@cpan.org>"
      ],
      "dist_name" => "Log-Dispatch-File-Stamped",
      "dist_version" => "0.15",
      "license" => "perl",
      "module_name" => "Log::Dispatch::File::Stamped",
      "recommends" => {},
      "recursive_test_files" => 1,
      "requires" => {
        "File::Basename" => 0,
        "File::Spec::Functions" => 0,
        "Log::Dispatch::File" => "2.38",
        "POSIX" => 0,
        "Params::Validate" => 0,
        "namespace::clean" => 0,
        "parent" => 0,
        "perl" => "5.006",
        "strict" => 0,
        "warnings" => 0
      },
      "script_files" => [],
      "test_requires" => {
        "ExtUtils::MakeMaker" => 0,
        "File::Spec" => 0,
        "Log::Dispatch" => 0,
        "Path::Tiny" => 0,
        "Test::Deep" => 0,
        "Test::More" => "0.88",
        "perl" => "5.006"
      }
    );


    my %fallback_build_requires = (
      "ExtUtils::MakeMaker" => 0,
      "File::Spec" => 0,
      "Log::Dispatch" => 0,
      "Path::Tiny" => 0,
      "Test::Deep" => 0,
      "Test::More" => "0.88",
      "perl" => "5.006"
    );


    unless ( eval { Module::Build->VERSION(0.4004) } ) {
      delete $module_build_args{test_requires};
      $module_build_args{build_requires} = \%fallback_build_requires;
    }

    my $build = Module::Build->new(%module_build_args);

    $build->create_build_script;
}
